src/core/mesh/qgsmeshdatasetgroupstore.h

```diff
@@ -0,0 +1,212 @@
+
+#ifndef QGSMESHDATASETGROUPSTORE_H
+#define QGSMESHDATASETGROUPSTORE_H
+
+#define SIP_NO_FILE
+
+#include "qgsmeshdataprovider.h"
+#include "qgsmeshdataset.h"
+
+/**
+ * \ingroup core
+ *
+ * Class used to store and access memory dataset group
+ *
+ * \since QGIS 3.16
+ */
+class QgsMeshMemoryDatasetStore: public QgsMeshDatasetSourceInterface
+{
+  public:
+
+    //! Adds a dataset group
+    void addDatasetGroup( const QgsMeshMemoryDatasetGroup &datasetGroup );
+
+    //! Removes the dataset group with the local \a index
+    void removeDatasetGroup( int index );
+
+    //! Returns whether if the dataset groups have temporal capabilities (a least one dataset group with more than one dataset)
+    bool hasTemporalCapabilities() const;
+
+    //! Returns the relative times of the dataset index with \a index, returned value in miliseconds
+    quint64 datasetRelativeTime( QgsMeshDatasetIndex index );
+
+    int datasetGroupCount() const override;
+    int datasetCount( int groupIndex ) const override;
+    QgsMeshDatasetGroupMetadata datasetGroupMetadata( int groupIndex ) const override;
+    QgsMeshDatasetMetadata datasetMetadata( QgsMeshDatasetIndex index ) const override;
+    QgsMeshDatasetValue datasetValue( QgsMeshDatasetIndex index, int valueIndex ) const override;
+    QgsMeshDataBlock datasetValues( QgsMeshDatasetIndex index, int valueIndex, int count ) const override;
+    QgsMesh3dDataBlock dataset3dValues( QgsMeshDatasetIndex index, int faceIndex, int count ) const override;
+    bool isFaceActive( QgsMeshDatasetIndex index, int faceIndex ) const override;
+    QgsMeshDataBlock areFacesActive( QgsMeshDatasetIndex index, int faceIndex, int count ) const override;
+
+    //! Not implemented, always returns false
+    bool addDataset( const QString &uri ) override;
+
+    //! Not implemented, always returns empty list
+    QStringList extraDatasets() const override;
+
+    //! Not implemented, always returns true
+    bool persistDatasetGroup( const QString &outputFilePath,
+                              const QString &outputDriver,
+                              const QgsMeshDatasetGroupMetadata &meta,
+                              const QVector<QgsMeshDataBlock> &datasetValues,
+                              const QVector<QgsMeshDataBlock> &datasetActive,
+                              const QVector<double> &times ) override;
+
+    //! Not implemented, always returns true
+    bool persistDatasetGroup( const QString &outputFilePath,
+                              const QString &outputDriver,
+                              QgsMeshDatasetSourceInterface *source,
+                              int datasetGroupIndex ) override;
+
+
+  private:
+    QVector<QgsMeshMemoryDatasetGroup> mGroups;
+};
+
+/**
+ * \ingroup core
+ *
+ * Class used to store and access all the dataset groups related to a mesh layer
+ *
+ * \since QGIS 3.16
+ */
+class QgsMeshDatasetGroupStore: public QObject
```

Comment:
`QgsMeshMemoryDatasetStore` is derived from `QgsMeshDatasetSourceInterface`, `QgsMeshDatasetGroupStore` is derived from `QObject`